sdk/src/Core/Sampling/ServiceConnector.cs

```diff
@@ -53,61 +56,103 @@ public ServiceConnector(DaemonConfig daemonConfig, AmazonXRayClient xrayClient)
             }
             _daemonConfig = daemonConfig;
            
-            if (xrayClient == null)
+            if (xrayConfig == null)
             {
-                xrayClient = CreateXRayClient();
+                xrayConfig = CreateXRayConfig();
             }
 
-            _xrayClient = xrayClient;
+            _xrayConfig = xrayConfig;
         }
 
-        private AmazonXRayClient CreateXRayClient()
+        private XRayConfig CreateXRayConfig()
         {
-            _config.ServiceURL = $"http://{_daemonConfig.TCPEndpoint.Address}:{_daemonConfig.TCPEndpoint.Port}";
-            return new AmazonXRayClient(_credentials,_config);
+            var config = new XRayConfig();
+            config.ServiceURL = $"http://{_daemonConfig.TCPEndpoint.Address}:{_daemonConfig.TCPEndpoint.Port}";
+            return config;
         }
 
         private void RefreshEndPoint()
         {
             var serviceUrlCandidate = $"http://{_daemonConfig.TCPEndpoint.Address}:{_daemonConfig.TCPEndpoint.Port}";
-         
-            if (serviceUrlCandidate.Equals(_xrayClient.Config.ServiceURL)) return; // endpoint do not need refreshing
 
-            _config.ServiceURL = serviceUrlCandidate;
-            _xrayClient = new AmazonXRayClient(_credentials, _config);
-            _logger.DebugFormat($"ServiceConnector Endpoint refreshed to: {_xrayClient.Config.ServiceURL}");
+            if (serviceUrlCandidate.Equals(_xrayConfig.ServiceURL)) return; // endpoint do not need refreshing
+
+            _xrayConfig.ServiceURL = serviceUrlCandidate;
+            _logger.DebugFormat($"ServiceConnector Endpoint refreshed to: {_xrayConfig.ServiceURL}");
         }
 
         /// <summary>
-        /// Use X-Ray client to get the sampling rules
-        /// from X-Ray service.The call is proxied and signed by X-Ray Daemon.
+        /// Get the sampling rules from X-Ray service.The call is proxied and signed by X-Ray Daemon.
         /// </summary>
         /// <returns></returns>
-        public async Task<GetSamplingRulesResponse>  GetSamplingRules()
+        public async Task<GetSamplingRulesResponse> GetSamplingRules()
         {
-            List<SamplingRule> newRules = new List<SamplingRule>();
-            GetSamplingRulesRequest request = new GetSamplingRulesRequest();
-
-            Task<Model.GetSamplingRulesResponse> responseTask;
+            Task<WebResponse> responseTask;
             lock (_xrayClientLock)
             {
                 RefreshEndPoint();
-                responseTask = _xrayClient.GetSamplingRulesAsync(request);
+                responseTask = GetSamplingInfoAsync(_xrayConfig.ServiceURL + "/GetSamplingRules");
```

Comment:
Just getting the task here and doing the await outside of the lock defeats the purpose of the lock. The actual getting the response for the request will happen outside of the lock. So if the point of the lock was to not have multiple calls to the sevice url at the same time his won't work.

I assume this was done because you can't do an await with the lock keyword. Usually you get around that by directly using the .NET monitor class.